samples/UpgradeStepSample/UpgradeStepSample/AuthorsPropertyUpgradeStep.cs

```diff
@@ -0,0 +1,213 @@
+            {
+                _logger.LogDebug("No authors setting found in extension configuration. Authors will not be added to the project.");
+
+                // UpgradeInitializeResult includes a step status, string with details on the status, and a BuildBreakRisk property
+                // indicating to the user whether applying this step is likely to introduce build breaks into their project.
+                var result = new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, "No authors need added", BuildBreakRisk.None);
+                return Task.FromResult(result);
+            }
+
+            var existingAuthorsProperty = projectFile.GetPropertyValue(AuthorsPropertyName);
+            if (existingAuthorsProperty is null || !existingAuthorsProperty.Contains(_options.Authors, StringComparison.OrdinalIgnoreCase))
+            {
+                _logger.LogDebug("Existing authors property ({Authors}) does not contain expected authors ({ExpectedAuthors})", existingAuthorsProperty, _options.Authors);
+
+                // If a change is needed, InitializeImplAsync should return a result with UpgradeStepStatus.Incomplete, but not make the update
+                // until the upgrade step's ApplyImplAsync method is called.
+                var result = new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"Expected authors {_options.Authors} need added to the projects Authors property", BuildBreakRisk.None);
+                return Task.FromResult(result);
+            }
+            else
+            {
+                _logger.LogDebug("An existing Authors property exists with expected Authors. Skipping AuthorsPropertyUpgradeStep");
+                var result = new UpgradeStepInitializeResult(UpgradeStepStatus.Complete, "Expected authors are already listed in the project file", BuildBreakRisk.None);
+                return Task.FromResult(result);
+            }
+        }
+
+        /// <summary>
+        /// The ApplyImplAsync step is invoked when the user attempts to actually apply the upgrade step and should make
+        /// necessary changes to the upgrade context.
+        /// </summary>
+        /// <param name="context">The upgrade context to update.</param>
+        /// <param name="token">A cancellation token.</param>
+        /// <returns>A result indicating whether the upgrade step applied successfully or not.</returns>
+        protected override async Task<UpgradeStepApplyResult> ApplyImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (_options?.Authors is null)
+            {
+                throw new InvalidOperationException("AuthorsPropertyUpgradeStep should not be applied without a valid authors option set.");
+            }
+
+            // The IProjectFile abstraction exposes the csproj file.
```

Comment:
does it matter to be csproj? Or just project file? (thinking of the vb users here)

Reply:
Good point. Fixed.